### ISSUE TITLE
Group items by function in outfitters beyond big categories of Turrets, Systems and Power

### ISSUE THREAD

**Issue by Tadrix** [endless-sky/endless-sky#2166]
Don't know if this has been brought up:
Right now the game groups them inside big categories by alphabetic order exclusively (unless I missed some commits, but I went through them), which in my opinion generates mess, especially for larger outfitters.
It would be easier to use if the game grouped them by subclasses and displayed items belonging to same subclass adjacent to each other, preferably somehow sorted by functionality, if at all possible, like this:
Turrets: offense, point defense (anti-missile).
Systems: defense [shields and hull, (ECM) jammers], sensors, cloaking (if applicable), FTL systems/drive, modifications, cooling.
Power: batteries, solar panels, fuel cells (and biochem cells), radioisotopic generators, reactors.

**Comment by w0ntoan** [comment]
A potential problem would be multipurpose items, e.g. a Korath Systems Core.  Is it a shield? A hull repair? A battery? Would it appear under all three subcategories? If not, which one should it be classified under?

**Comment by Tadrix** [comment]
Korath Systems Core would be a shield (since it's already in Systems, not Power), and it should appear once, under shields. Similarly, Wayfarer engines are still engines and Wayfarer reactors are reactors, not batteries.

**Comment by Bladewood** [comment]
wanderer*

this feels incredibly arbitrary.

**Comment by Tadrix** [comment]
Yeah, sorry.

Does it? Some Korath turrets have power collection abilities, but no one is saying they should be under batteries, right? I'd say we look at item's primary function. Battery functions in items with other functions are likely to support its main functionality which draws power. Hull repair is the same thing as shield repair and there's no item which repairs hull exclusively (at least AFAIK), so that's not a problem, even if we ignore that it would still be the same thing (effectively hull + shields = still hit points, one regenerates one part, the second one regenerates the other part, we could thing of some third component and it still wouldn't change anything, still hit points).

**Comment by Amazinite** [comment]
>there's no item which repairs hull exclusively
Coalition repair modules.

**Comment by Tadrix** [comment]
Ok, then, they should just be in the same subgroup as shields. Anyway, it gives the player opportunity to seek just in couple of places, as opposed to A-Z.

**Comment by MessyMix** [comment]
>Wayfarer engines.

Anyways. Not sure what this really is going to achieve.

What I _can_ see happening right now is a split of just "systems" into perhaps "Protection", "Batteries", and "Cooling" (perhaps more that I'm missing), but I don't see how this would make the game _so_ much better. It's too much hassle for too little reward, imo.

And as for the Turret subcategory,
no.

**Comment by Amazinite** [comment]
How many outfitters are really so big that it's difficult for you to find what you're looking for within the current categories?

**Comment by Tadrix** [comment]
@MessyMix If these would all be visible categories, than that would be a mess, I agree. My point in OP was to just group them together (so, for example, that there's still "turrets" category, in which game first lists all turrets and then all AM's, there's still "power", but batteries are always listed as first in the list, etc.). If needed, I can try to make a mock up for how I originally envisaged it.

@Amazinite More of a situation — you're going to make some massive upgrades, so you head to the biggest outfitter (most likely something convenient in human space, Valhalla, Wayfarer or Zug, for example). You are hauling a lot of just looted outfits and those you intend to put on your ship just after you'll finely swap out that reactor and gain some space. Additionally, you plan to sell some of your ships, and in order to gain funds you do it first, with all of their alien outfits… do you see where I'm heading? Once there's dozens of different outfits "in stock", the outfitter becomes a bloody mess, and one cannot just take off and land, because there might still be some needed ones.

**Comment by DingusShingleton** [comment]
Personally I have always thought that the outfitter could use some improvements.... After hundreds of hours, the initial impression fades mightily, but I still remember my first few hours of gameplay and how counter intuitive the display of outfits seemed to me. At this point it's second nature so changing it seems pointless, but I think in these matters taking heed of the thoughts of newer players is never a bad idea.

**Comment by KiLEdEnNis** [comment]
> How many outfitters are really so big that it's difficult for you to find what you're looking for within the current categories?

my custom ones that contain everything (including my mod) on Deep & Spera Anatrusk ;-;

what CAN be added is a filter (and enabling multiple filters at once)
filter to outfits that have cooling, or ones that can store energy, or repair hull, or both energy storage & hull
weapons that do heat damage, or weapons that do ion damage
it'd mostly be for the systems category (and a bit for the power category), but they're the biggest ones, so it would really make it easier to for example compare outfits or quickly see what items have for example cooling if you need some

**Comment by Amazinite** [comment]
How about just having the find function work in outfitters/shipyards similarly to how it works on the map?

**Comment by jafdy** [comment]
I would support splitting Systems up in to 2/3 categories but splitting things up too much risks adding more clutter then you remove.

**Comment by Amazinite** [comment]
>More of a situation

Even in a situation like that, it's still all alphabetical, so as long as you know your ABCs and what you're looking for it isn't usually that much of a problem. I usually just find stuff by its outfit image anyways.

**Comment by Tadrix** [comment]
@Amazinite It only works when you know specifically what you're looking for, not when you only know what kind of thing you're looking for. @KiLEdEnNis has an excellent idea.

Also, let player sort items inside categories. Some ideas: by race, by outfit space, by power capacity / outfit space, by (range * shield damage) / oufit space, and such effectiveness things (perhaps @endless-sky himself would know better which things are most relevant, as evidenced in  #1771). And make game remember last sorting priorities for each category. I think then everybody would have exactly what they want.

**Comment by Amazinite** [comment]
(Range * shield damage) / outfit space? Slow down. At that point you're asking the game to point you to what the best stuff is. Equations like that are what people use in spreadsheets that they make outside of the game when they want to seriously min/max their outfits. If sorting were a thing, I would saw sort by cost, by mass or outfit space, and by the main function of an outfit (e.g. Power capacity, energy generation, thrust, steering, etc.), but if you want to find out the best stuff by doing things like ln((range * (shield dmg/s + hull dmg/s))*100) then do that on a spreadsheet. Having that in game invokes an "automation of mechanics" arguments.

**Comment by KiLEdEnNis** [comment]
@Tadrix I had the idea of "by race", but try and name 1 outfitter that has outfits of more than 1 race (Coalition+Heliarch count as 1)
adding the option is fine for modders but not really needed

and an addition to my earlier suggestion, being able to sort outfits by space instead of alphabetical (as outfit space is a universal stat) seems good too

**Comment by Wrzlprnft** [comment]
i'd judt make some filters, with some stats a putfit can have. those attributes which are not ticked, the outfits only having those dissappear. one could look precisely for shield regeneration, thrust, steering, energy generation;storage

**Comment by gunqqer** [comment]
as someone who uses world forge a lot, some sort of system other than ABCs would be nice. And I can only go so far with just images due to a lot of plugins I have don't have images for the outfits.

**Comment by Tadrix** [comment]
Ok, I wasn't really trying to make spreadsheets redundant, since this anyway wouldn't replace info given by a spreadsheet, but I understand the point. How about, for example, outfit space taken (since it's most important thing when trying to squeeze outfits in a ship), or race and then outfit space taken, or type (for engines) and then outfit space taken, etc.?

The reason I'm mentioning races or types is player's own ships can carry many outfits from various sources and it can be a real mess in the outfitter (because these are shown, and even worse this situation is after long trip of pludering). Then, for example, if I arrive to a human outfitter and try to compare my Hai Gorge Batteries to their LP144a Battery Pack (which is closest by space taken), it would really help if these were nearby in the list instead of being several lines away. Same with plundered Generator (Candle Class) and Small Biochemical Cell, etc. What I proposed in my OP was that these would be invisibly subgrouped (in line with others, just to be adjacent, without visible division to subcategories which would lead to one or two outfits in a such), and sorting by race (but sometimes grouping human and Hai tech together because of similarities), in other words, putting together by function and then sorting by outfit space.

Examples (no Coalition tech in examples, I haven't interacted with them much yet, and some things might be missing, it's only an example):

> Guns by reload (ascending) and then mass;
>
> Turrets
> by required crew (descending):
> offensive (>0 crew, by reload (ascending) and then mass);
> anti-missile (0 crew, by range and then mass);
>
> Secondary Weapons by homing and then mass (both ascending);
>
> Ammunition by mass and then cost (both ascending);
>
> Systems
> defense: group shields by race and then size: human, Hai, Wanderer, Korath, Quarg; then Small Radar Jammer and Large Radar Jammer;
> then sensors (Cargo Scanner, Outfit Scanner, Surveillance Pod); then cloaking (Cloaking Device);
> FTL systems/drive would be listed together (first ramscoops, by race and outfit space, then drives, by how advanced they are): Ramscoop, Catalytic Ramscoop, Wanderer Ramscoop, Hyperdrive, Scram Drive, Jump Drive;
> then modifications (expansions by o/s, bunks by abc, other by abc): Mass Expansion, Cargo Expansion, Bunk Room, Small Bunk Room, Fuel Pod, Interference Plating;
> human and Hai cooling would be listed together and in outfit space order, then Wanderer and then Korath tech: Cooling Ducts, Hai Williwaw Cooling, Water Coolant System, Liquid Nitrogen Cooler, Liquid Helium Cooler, Wanderer Heat Sink, Small Heat Shunt, Large Heat Shunt (or, if it's not possible, just by outfit space);
>
> Hand to Hand by capture attack (descending) and then capture defense (ascending): Nerve Gas, Fragmentation Grenades, Laser Rifle, Security Station, Intrusion Countermeasures, Pug Biodefenses;
>
> Power
> batteries by outfit space: Supercapacitor, Hai Chasm Batteries, LP036a Battery Pack, Hai Fissure Batteries, LP072a Battery Pack, Hai Gorge Batteries, LP144a Battery Pack, Nanotech Battery, Hai Ravine Batteries, LP288a Battery Pack, Hai Valley Batteries, LP576a Battery Pack; 
> solar panels by outfit space: KP-6 Photovoltaic Panel and KP-6 Photovoltaic Array;
> cells by outfit space: Small Biochemical Cell, nGVF-AA Fuel Cell, Sand Cell, nGVF-BB Fuel Cell, nGVF-CC Fuel Cell, nGVF-DD Fuel Cell, Large Biochemical Cell, nGVF-EE Fuel Cell;
> radioisotopic generators by outfit space: Generator (Candle Class), Generator (Furnace Class), RT-I Radiothermal, S3 Thermionic, NT-200 Nucleovoltaic, Generator (Inferno Class);
> reactors (not incl. Wanderer) by outfit space: Pebble Core, Dwarf Core, Plasma Core, Geode Reactor, Fission Reactor, Breeder Reactor, Boulder Reactor, Double Plasma Core, Fusion Reactor, Armageddon Core, Stack Core, Triple Plasma Core;
> combined (reactors+batteries) by outfit space: Red Sun Reactor, Yellow Sun Reactor, White Sun Reactor, Blue Sun Reactor;
>
> Engines by type (flare sprite) and then outfit space: afterburners, graviton, ion (incl. Pug), Wanderer, plasma, atomic (incl. Hai), Korath.

**Comment by KiLEdEnNis** [comment]
so much words ;-;
not to be rude but people sometimes don't bother reading huge word walls

**Comment by Tadrix** [comment]
Not anything intuitively understood is understandably explainable. Or I'm just not good at this. Probably easier would be write it as a code. ><
But since I'm not good at C++ either, oh, well, take 2:

**TL;DR version:** 
Items could be grouped by origin (race) and outfit space. Group them in line with others, not as visible subcategories but just to be adjacent with each other.
More elaborate grouping (**asc**ending, **desc**ending, o/s — **o**utfit **s**pace, other explanations in a previous post, if needed, no Coalition in examples):
**Guns**: reload (asc), mass (asc);
**Turrets**: offensive — req. crew (desc), reload (asc), mass (asc); anti-missile — range (asc), mass (asc);
**Ammunition**: homing (asc), mass (asc);
**Systems**:
defense: shields — race [human, Hai, Wanderer, Korath, Quarg], o/s (asc); radar jammers — o/s (asc);
sensors; cloaking;
FTL: ramscoops — race [human, Wanderer], o/s (asc); drives — [_Hyperdrive_, _Scram Drive_, _Jump Drive_];
mods: expansions — o/s (asc); bunks — abc (asc); other — abc (asc);
cooling: race [human&Hai, Wanderer, Korath], o/s (asc) (or just by asc o/s);
**Hand to Hand**: capture attack (desc), capture defense (asc);
**Power**: type [batteries, solar panels, cells, radioisotopic generators, reactors, combined (reactor+battery)], o/s (asc);
**Engines**: type (flare sprite) [afterburner, graviton, ion, Wanderer, plasma, atomic, Korath], o/s (asc).

Nah, it's too long either. Ok, I'll go with…
____________________

**ADHD version:** Group items mainly by outfit space occupied.

**Comment by KiLEdEnNis** [comment]
sorting by species isn't needed for the base game; name 1 outfitter where you see outfits from over 1 "species", although I'd like it for mods (like world forge)
giving each tab the option to sort alphabetically or by size, or certain tags (energy generation, cooling) would be a great addition & I think that's basically your request?

**Comment by MessyMix** [comment]
What do people think of a separate section for "installed" items? I'm not sure how I would feel about it; would an entirely new section just for installed outfits help? I feel that it might be beneficial for managing, but at the same time, it may get complicated.

**Comment by Tadrix** [comment]
@KiLEdEnNis Any outfitter with player's ships in it. Also, any outfitter player sold outfits at. Besides, I'd expect some human outfits appearing in Hai space — humans mingle with Hai long enough, and Hai could be interested in human beams (marginally), sensors (marginally), ramscoops (these and beams perhaps warranting a mission to Deep Sky), fuel pods and cooling systems (especially these, warranting a mission to Kraz Cybernetics), opening up interesting trade possibilities (missions to bring these outfits to Hai space? also, perhaps some missions where player gets import/export business going, which makes these things available in Hai space and generates player an income). How does it sound?

@MessyMix sounds good, but what about installed items which also are offered at particular outfitter? Would they show up in both sections? Also, this doesn't make the job of comparing player's outfits to offered ones easier.

**Comment by KiLEdEnNis** [comment]
a side note: fuel pods are "universal": all species have the same one
but adding a tab for Hai, Coalition, Wanderer, Korath & Quarg outfits because you have 1 outfit from each species installed looks bad

**Comment by Tadrix** [comment]
No tabs. Grouped together in line. As an example, situation now:
**Category**
[A Human item] [B Hai item] [C Wanderer item] [D Human item] [E Hai item] [F Human item]
A better one:
**Category**
[A Human item] [D Human item] [F Human item] [B Hai item] [E Hai item] [C Wanderer item]
No additional tabs, no additional categories, just grouped (sorted) together in the same line.

**Comment by KiLEdEnNis** [comment]
Kor Efret space: fuel pods
Wanderer space: fuel pods
about that grouping thing, an easy way is to add the species name before outfits, but since that'd break save files that isn't an option, + it would look bad
it's only alphabetical now, if there would be filters like species ("universal stuff" like the fuel pod, hyperdrive etc. would go first) or by size this'll be solved

**Comment by Tadrix** [comment]
Yup.

@KiLEdEnNis Hai space: no fuel pods. I figure it would be logical for Hai to buy from humans (they don't seem to manufacture their own fuel pods), because,
first, they don't need jump drive and fly through Unfettered space to reach them (there's a ??? wormhole),
second, there are already human merchants in Hai space and diplomatic relations with them are century old,
third, Kor Mereti are robotic enemies, which makes trading unlikely (unless friendly squirrels resolve to just haul entire Kor Mereti factory to Hai space and reprogram it, which sounds like fun),
fourth, humans (seem to) possess greatest exporting capacity and willingness to trade outfits.

**Comment by KiLEdEnNis** [comment]
the Hai doesn't really need fuel pods, but I get why ya would want them to sell them
